spec.md

```diff
@@ -499,6 +499,56 @@ the module.  The `c`{.firrtl} sub-field contained in the `b`{.firrtl} sub-field
 flows into the module, and the `d`{.firrtl} sub-field contained in the
 `b`{.firrtl} sub-field flows out of the module.
 
+## Type Modifiers
+
+### Constant Type
+
+A constant type is a type whose value is guaranteed to be unchanging at circuit 
+execution time.  Constant is a constraint on the mutability of the value, it 
+does not imply a literal value at a point in the emitted design.  Constant types 
+may be used in ports, wire, nodes, and generally anywhere a non-constant type is 
+usable.  Operations on constant type are well defined.  As a general rule (with 
+any exception listed in the definition for such operations as have exceptions), 
+an operation whose arguments are constant produces a constant.  An operations 
+with some non-constant arguments produce a non-constant.  Constants can be used 
+in any context with a source flow which allows a non-constant.  Constants may be 
+used as the target of a connect so long as the source of the connect is itself 
+constant.  These rules ensure all constants are derived from literals or from 
+constant-typed input ports of the top-level module.
+
+``` firrtl
+const UInt<3>
+const SInt
+const {real: UInt<32>, imag : UInt<32>, other : const SInt}
+const 
+```
+
+Last-connect semantics of constant typed values is well defined, so long as any 
+control flow is conditioned on an expression which has a constant type.  This 
+means if a constant is being assigned to in a `when` block, the `when`'s 
+condition must be a constant.
+
+Output ports of external modules and input ports to the top-level module may be
+constant.  In such case, the value of the port is not known, but that it is 
+non-mutating at runtime is known.
+
+The indexing of a constant aggregate produces a constant of the appropriate type 
+for the element.
+
+#### A note on implementation
+
+Constant types are a restriction on firrtl types.  Therefore, firrtl structures 
+which would be expected to produce certain verilog structures will produce the
+same structure if instantiated with a constant type.  For example, an input port
+of type `const UInt` will result in a port in the verilog, if under the same
+conditions an input port of type `UInt` would have.
```

Comment:
Hmm, how do we ensure this? I know it's an implementation detail but is it now an assertion/guarantee that no optimization or composition of them cannot discover a port to be dead due to const-ness that would not otherwise?

Don't have a complete example, but seems possible to have something like: a register initialized+assigned from a constant input that we could optimize to that input value when we know it's constant but cannot if the value may change (where we need to dutifully store the latest value at clock edge or so).  Or maybe just the "has this changed since initialization" sort of module where the output is a comparison of the register to the input signal it's initialized with?